File: vantix_back/app/api/v1/controller/crm.py
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, File, UploadFile, Form
from sqlalchemy.orm import Session
from app import crud, models, schemas
from app.api import deps
from app.services.gamificacion.kpi_service import kpi_service
from app.services.common.file_manager import FileManager

router = APIRouter()

# --- LLAMADAS ---

@router.post("/llamadas/", response_model=schemas.crm.LlamadaResponse)
async def registrar_llamada(
    *,
    db: Session = Depends(deps.get_db),
    current_user: models.empleado.Empleado = Depends(deps.get_current_active_user),
    id_plan: int = Form(...),
    numero_destino: str = Form(...),
    resultado: str = Form(...),
    nombre_destinatario: Optional[str] = Form(None),
    duracion_segundos: int = Form(0),
    notas_llamada: Optional[str] = Form(None),
    foto_prueba: Optional[UploadFile] = File(None)
):
    """
    Registra una llamada realizada y suma 1 punto al KPI.
    Permite subir una foto opcional como prueba.
    """
    # 1. Validar Plan
    plan = crud.plan.get(db, id=id_plan)
    if not plan:
        raise HTTPException(status_code=404, detail="Plan no encontrado")
    
    # 2. Gestionar Foto si existe
    url_foto = None
    if foto_prueba:
        url_foto = await FileManager.save_upload_file(foto_prueba, subdirectory="crm/llamadas", prefix="call")

    # 3. Preparar data
    llamada_in = schemas.crm.LlamadaCreate(
        id_plan=id_plan,
        numero_destino=numero_destino,
        resultado=resultado,
        nombre_destinatario=nombre_destinatario,
        duracion_segundos=duracion_segundos,
        notas_llamada=notas_llamada,
        url_foto_prueba=url_foto
    )
    
    # 4. Guardar en DB
    db_obj = crud.llamada.create(db, obj_in=llamada_in)
    
    # 5. Actualizar KPI (1 punto por llamada)
    kpi_service.update_kpi_metrics(
        db, 
        id_plan=id_plan, 
        field="real_llamadas", 
        puntos=1
    )
    
    return db_obj

@router.get("/llamadas/", response_model=List[schemas.crm.LlamadaResponse])
def listar_llamadas(
    db: Session = Depends(deps.get_db),
    current_user: models.empleado.Empleado = Depends(deps.get_current_active_user),
    id_plan: Optional[int] = None,
    skip: int = 0,
    limit: int = 100
):
    if id_plan:
        return crud.llamada.get_multi_by_plan(db, id_plan=id_plan, skip=skip, limit=limit)
    return crud.llamada.get_multi(db, skip=skip, limit=limit)


# --- EMAILS ---

@router.post("/emails/", response_model=schemas.crm.EmailResponse)
async def registrar_email(
    *,
    db: Session = Depends(deps.get_db),
    current_user: models.empleado.Empleado = Depends(deps.get_current_active_user),
    id_plan: int = Form(...),
    email_destino: str = Form(...),
    asunto: Optional[str] = Form(None),
    estado_envio: str = Form("Enviado"),
    foto_prueba: Optional[UploadFile] = File(None)
):
    """
    Registra un email enviado y suma 1 punto al KPI.
    Permite subir una foto opcional (captura de pantalla) como prueba.
    """
    plan = crud.plan.get(db, id=id_plan)
    if not plan:
        raise HTTPException(status_code=404, detail="Plan no encontrado")
    
    url_foto = None
    if foto_prueba:
        url_foto = await FileManager.save_upload_file(foto_prueba, subdirectory="crm/emails", prefix="email")

    email_in = schemas.crm.EmailCreate(
        id_plan=id_plan,
        email_destino=email_destino,
        asunto=asunto,
        estado_envio=estado_envio,
        url_foto_prueba=url_foto
    )

    db_obj = crud.email.create(db, obj_in=email_in)
    
    # 1 punto por email
    kpi_service.update_kpi_metrics(
        db, 
        id_plan=id_plan, 
        field="real_emails", 
        puntos=1 
    )
    
    return db_obj

@router.get("/emails/", response_model=List[schemas.crm.EmailResponse])
def listar_emails(
    db: Session = Depends(deps.get_db),
    current_user: models.empleado.Empleado = Depends(deps.get_current_active_user),
    id_plan: Optional[int] = None,
    skip: int = 0,
    limit: int = 100
):
    if id_plan:
        return crud.email.get_multi_by_plan(db, id_plan=id_plan, skip=skip, limit=limit)
    return crud.email.get_multi(db, skip=skip, limit=limit)

File: vantix_mobile/src/api/crm.ts
import { apiClient } from './client';

export interface LlamadaData {
    id_plan: string;
    numero_destino: string;
    nombre_destinatario?: string;
    duracion_segundos?: string;
    resultado: string;
    notas_llamada?: string;
    foto_prueba?: string;
}

export interface EmailData {
    id_plan: string;
    email_destino: string;
    asunto?: string;
    estado_envio: string;
    foto_prueba?: string;
}

export const crmService = {
    registrarLlamada: async (data: LlamadaData) => {
        const formData = new FormData();
        formData.append('id_plan', data.id_plan);
        formData.append('numero_destino', data.numero_destino);
        formData.append('resultado', data.resultado);
        if (data.nombre_destinatario) formData.append('nombre_destinatario', data.nombre_destinatario);
        if (data.duracion_segundos) formData.append('duracion_segundos', data.duracion_segundos);
        if (data.notas_llamada) formData.append('notas_llamada', data.notas_llamada);

        if (data.foto_prueba) {
            const filename = data.foto_prueba.split('/').pop();
            const match = /\.(\w+)$/.exec(filename || '');
            const type = match ? `image/${match[1]}` : `image`;
            formData.append('foto_prueba', { uri: data.foto_prueba, name: filename, type } as any);
        }

        const response = await apiClient.post('/api/v1/crm/llamadas/', formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        });

        return response.data;
    },

    registrarEmail: async (data: EmailData) => {
        const formData = new FormData();
        formData.append('id_plan', data.id_plan);
        formData.append('email_destino', data.email_destino);
        if (data.asunto) formData.append('asunto', data.asunto);
        if (data.estado_envio) formData.append('estado_envio', data.estado_envio);

        if (data.foto_prueba) {
            const filename = data.foto_prueba.split('/').pop();
            const match = /\.(\w+)$/.exec(filename || '');
            const type = match ? `image/${match[1]}` : `image`;
            formData.append('foto_prueba', { uri: data.foto_prueba, name: filename, type } as any);
        }

        const response = await apiClient.post('/api/v1/crm/emails/', formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        });

        return response.data;
    },
};
